numpy/core/src/multiarray/item_selection.c

```diff
@@ -2093,13 +2093,17 @@ PyArray_CountNonzero(PyArrayObject *self)
         PyArray_PREPARE_TRIVIAL_ITERATION(self, count, data, stride);
 
         while (count--) {
-            if (nonzero(data, self)) {
+            npy_bool n = nonzero(data, self);
+            if (n == -1) {
```

Comment:
I think a lot of the ufunc inner loops had a similar problem which was solved by adding a `needs_api` variable to signify whether the error check could be delayed until after the loop.